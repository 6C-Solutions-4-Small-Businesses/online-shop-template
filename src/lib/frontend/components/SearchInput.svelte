<script lang="ts">
    import SearchIcon from '$lib/frontend/components/icons/SearchIcon.svelte'

    export let inputId: string
    export let buttonId: string
    export let width: string | undefined = undefined
    export let onSearchSubmitHandler: ((searchTerm: string) => Promise<void>) | undefined = undefined
    export let onSearchTermChangeHandler: ((searchTerm: string) => void) | undefined = undefined
    export let onSearchResetHandler: (() => void) | undefined = undefined

    export let searchTerm: string = ''
    let isSearchTermInvalid: boolean = false

    $: isSearchTermInvalid = !searchTerm || searchTerm === "" || searchTerm.length <= 1

    async function onEnterPressSubmitSearch(event: KeyboardEvent): Promise<void> {
        if (event.code === "Enter") {
            await onSearchSubmitHandler?.call({}, searchTerm)
        }
    }

    function resetInput(): void {
        searchTerm = ''
        onSearchResetHandler?.call({})
    }

    function onSearchInputChangeHandler(event: Event): void {
        onSearchTermChangeHandler?.call({}, searchTerm)
    }

</script>

<div class={`${width} h-10 flex`}>
    <div class="{`relative w-full lg:w-8/12 lg:mb-0 mb-4`}">
        <input
                data-testid={inputId}
                bind:value={searchTerm}
                class="w-full h-12 focus:outline-transparent text-black text-left pl-2 rounded-l-lg border-none"
                on:keypress={onEnterPressSubmitSearch}
                on:input={onSearchInputChangeHandler}
                placeholder="Chercher un produit"
                type="text"
        >
        {#if searchTerm}
            <button
                    data-testid="search-input-clear-button"
                    class="absolute w-6 h-6 flex items-center justify-center text-sm text-white right-3 bottom-2 border rounded-full bg-orange-500 pb-0.5 shadow-md"
                    on:click={resetInput}
            >x
            </button>
        {/if}
    </div>
    <button
            data-testid={buttonId}
            class='h-12 text-sm xl:text-lg text-white p-1 rounded-r-lg bg-white'
            disabled={isSearchTermInvalid}
            on:click={() => onSearchSubmitHandler?.call({}, searchTerm)}
    >
        <SearchIcon classNames={`flex items-center rounded-r-lg py-1 px-2.5 h-10 text-black ${isSearchTermInvalid ? "bg-primary": "bg-teal-950"} p-1`}/>
    </button>
</div>
