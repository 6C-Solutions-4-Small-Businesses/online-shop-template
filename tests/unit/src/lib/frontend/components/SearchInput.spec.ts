import SearchInput from '$lib/frontend/components/SearchInput.svelte';
import '@testing-library/jest-dom';
import { cleanup, fireEvent, render, type RenderResult, screen } from '@testing-library/svelte';
import { afterEach, beforeEach, describe, expect } from 'vitest';

describe('Search input', () => {
	let renderResult: RenderResult<SearchInput, typeof import('@testing-library/dom/types/queries')>;
	const inputId = 'search-input-id';
	const buttonId = 'search-input-button-id';

	beforeEach(() => {
		renderResult = render(SearchInput, {
			props: {
				inputId,
				buttonId,
				width: 'w-20'
			}
		});
	});

	it('should have match snapshot', () => {
		expect(renderResult.container).toMatchSnapshot();
	});

	it('Search button should not be disable when searched product name length is greater than two', async () => {
		const searchButton = screen.getByTestId(buttonId) as HTMLButtonElement;
		const searchInput = screen.getByTestId(inputId) as HTMLButtonElement;

		await fireEvent.input(searchInput, { target: { value: 'riz' } });

		expect(searchButton.disabled).toBe(false);
	});

	it('Search button should be disable when searched product name length is inferior to two', async () => {
		const searchButton = screen.getByTestId(buttonId) as HTMLButtonElement;
		const searchInput = screen.getByTestId(inputId) as HTMLButtonElement;

		await fireEvent.input(searchInput, { target: { value: 'r' } });

		expect(searchButton.disabled).toBe(true);
	});

	afterEach(() => {
		cleanup();
		vi.clearAllMocks();
	});
});
