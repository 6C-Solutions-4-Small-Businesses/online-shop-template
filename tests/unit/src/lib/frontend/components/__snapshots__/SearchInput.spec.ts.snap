// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search input > should have match snapshot 1`] = `
<div>
  <div
    class="w-20 h-10 flex justify-center"
  >
    <div
      class="relative w-8/12"
    >
      <input
        class="w-full h-full focus:outline-none text-black text-left pl-2 rounded-none border border-l-gray-500 border-t-gray-500 border-b-gray-500"
        data-testid="search-input-id"
        placeholder="Chercher un produit"
        type="text"
      />
       
    </div>
     
    <button
      class="w-3/12 h-full text-sm xl:text-lg text-white bg-slate-500 p-1"
      data-testid="search-input-button-id"
      disabled=""
    >
      Rechercher
    </button>
  </div>
</div>
`;
